### ISSUE TITLE
Left to right rendering effect

### ISSUE THREAD

**Issue by mjip** [peterbrittain/asciimatics#263]
**Is your feature request related to a problem? Please describe.**
I would like for an effect to render text (or ASCII characters) from left to right, like it's being typed out.

**Describe the solution you'd like**
Similar to the mirage effect (where text can be rendered one character at a time, only randomly), it would be cool to have text rendered one letter at a time from left to right, so it appears to be typed in real-time.

**Comment by peterbrittain** [comment]
Makes sense.  I get the idea...  Would you be interested in implementing it?  Should be relatively easy to use `Mirage` as a template for the new `Effect` and change the logic to maintain a count of where you've got to in the text to be displayed and just print up to that (then increment the counter for the next update)...

**Comment by mjip** [comment]
Yep, PR incoming :^) It was relatively easy using the `Mirage` class as a template

**Comment by vinayak-mehta** [comment]
I was able to [implement this](https://github.com/vinayak-mehta/present/pull/20/files#diff-851333edd38d48b1cdacb13bf0007528R37) by extending `DynamicRenderer`. I'm passing in the text I want to be printed one character at a time, and saving the state of how much of that text has been consumed each time `DynamicRenderer` is called. This could be another way of adding the `Typewriter` effect to asciimatics, I would love to see this in the `basics.py` example! It looks like this:

![gif](https://vinayak.io/files/codio.gif)

The API to use this class looks similar to this:

```
Print(
    screen,
    Typewriter(text="import os", width=100, height=100),
    y,
    x=x,
    colour=Screen.COLOUR_WHITE,
    bg=Screen.COLOUR_BLACK,
    transparent=False,
    speed=4,
)
```

@peterbrittain What do you think about the implementation? I can work on creating a PR based on the fixes / API changes that you suggest.

**Comment by peterbrittain** [comment]
Thanks @vinayak-mehta .  That looks great!  Zorry for the delay...  I've been thinking about the best way to handle these things.  

After much musing, I think that it makes most sense for this to be a chained renderer that takes another renderer as input and gradually displays the content of that renderer using similar logic to your existing renderer.  Key benefit is that this allows more composition (e.g. for another renderer that supports coloured text, which could be the existing `StaticRenderer` right now or a future syntax highlighting renderer for your code example).

This slightly complicates  your simple case (just type some pre-defined text), but only by adding one more function call.  For example:

```
Print(
    screen,
    Typewriter(text=StaticRenderer(images=["import os"]), width=100, height=100),
    y
)
```

What do you think?  Is that worth the change to your code?

BTW - I don't think we should worry about FigletText.  Frankly, doing that a letter at a time will require detailed knowledge of the figlet font and so will be a separate custom renderer (which we can worry about another day).

**Comment by vinayak-mehta** [comment]
@peterbrittain Sorry for the late reply, this week has been busy but I'm planning to work on this over the weekend!

> After much musing, I think that it makes most sense for this to be a chained renderer that takes another renderer as input and gradually displays the content of that renderer using similar logic to your existing renderer. Key benefit is that this allows more composition (e.g. for another renderer that supports coloured text, which could be the existing StaticRenderer right now or a future syntax highlighting renderer for your code example).

This is a great suggestion and makes a lot of sense, I hadn't thought about this. A future syntax highlightning renderer would be awesome! Can you give me some pointers on how to make something like syntax highlighting work? Right now I'm doing the `DynamicRenderer`'s `_write()` method to do something like self._write(text, x, y, colour=Screen.COLOUR_GREEN). Would syntax highlighting involve doing multiple `self._write` calls for different parts of the code at different coordinates with different colors? Is there another way?

> What do you think? Is that worth the change to your code?

The suggestion makes a lot of sense. I guess instead of extending `DynamicRenderer` in the implementation I mention above, I could just use `Typewriter` and have a list of prints with _static_ and _dynamic_ prints to simulate a terminal session playback.

> BTW - I don't think we should worry about FigletText. Frankly, doing that a letter at a time will require detailed knowledge of the figlet font and so will be a separate custom renderer (which we can worry about another day).

:+1:

**Comment by peterbrittain** [comment]
@vinayak-mehta I'm not quite sure what you're getting at for possible implementation...  Let me try to elucidate what I meant.

Have a look at the `Rainbow` renderer and how it is used.  You could use something similar for a `Typewriter` renderer.  Instead of adding volours, you could use the full text and colour map and just return an extra character each time.  Using a dynamic renderer that writes directly to the buffer would be more memory efficient and so better.

To create a syntax colouring option, I suggest we create a renderer that can use a parser to convert a blob of text into text/colour map.  The next step would be to create a pygments parser to use in that renderer.

**Comment by vinayak-mehta** [comment]
Thanks for the pointer! I'll have to look into the Rainbow renderer and learn how color maps work. I took a look at the `xmas.py` example to see how the [${c,a,b}](https://github.com/peterbrittain/asciimatics/blob/master/asciimatics/renderers.py#L131) syntax works and it makes sense to me now! I didn't know about this syntax before. I was able to use it to add text formatting support to [present](https://github.com/vinayak-mehta/present)!

I'll try to work on the `Typewriter` PR today.

On a separate note, would you like a PR to add a `Screen.A_ITALIC` attr (and maybe also more attrs)? I saw that [Python 3.7 added A_ITALIC in the curses module](https://docs.python.org/3/library/curses.html#curses.ncurses_version). I guess this change would have to be added in a backwards compatible way.

**Comment by peterbrittain** [comment]
Thanks for the offer, but I don't think we should add more attributes.  I've tried to keep asciimatics limited to stuff that you can do in all environments and deliberately avoided the less well supported ones.  In the case for italics, it is only possible to do that in recent builds of Windows 10 (that support ansi escape codes).  It might make sense to try that out once I have added support for that screen type, but not just yet.

**Comment by vinayak-mehta** [comment]
Makes sense :+1: 

Sorry for being inactive on the implementation. I've been busy with some other deadlines + a medical emergency at home. I will get back to this as soon as I can find time.

**Comment by peterbrittain** [comment]
NP.  Real life always comes first...

**Comment by peterbrittain** [comment]
This issue has various ideas in it, which are still worth doing at some point.  For now I just wanted to flag that the AnsiArtPlayer and AsciinemaPlayer renderers may do what you want in some cases.

**Comment by peterbrittain** [comment]
Had some spare time...  This is now available as of f63832aa370ac2e4a5d54273d1f01a4a087d77d9
